File: app/src/main/java/com/ne1c/gitteroid/presenters/LoginPresenter.java
package com.ne1c.gitteroid.presenters;

import com.ne1c.gitteroid.R;
import com.ne1c.gitteroid.dataproviders.DataManger;
import com.ne1c.gitteroid.ui.views.LoginView;
import com.ne1c.gitteroid.utils.RxSchedulersFactory;
import com.ne1c.gitteroid.utils.Utils;

import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

public class LoginPresenter extends BasePresenter<LoginView> {
    private final String CLIENT_ID = "247736d87aa0134a33f73b00cc47b18165296e9e";
    private final String CLIENT_SECRET = "238ca926e7e59121ceb7c46c3a2c91eafe51b6c5";
    private final String GRANT_TYPE = "authorization_code";
    private final String REDIRECT_URL = "http://about:blank";
    private final String RESPONSE_TYPE = "code";

    public final String AUTH_URL = "https://gitter.im/login/oauth/authorize?"
            + "client_id=" + CLIENT_ID + "&response_type=" + RESPONSE_TYPE + "&redirect_uri=" + REDIRECT_URL;

    private LoginView mView;

    private CompositeSubscription mSubscriptions;
    private DataManger mDataManager;
    private RxSchedulersFactory mFactory;

    public LoginPresenter(RxSchedulersFactory factory, DataManger dataManager) {
        mDataManager = dataManager;
        mFactory = factory;
    }

    @Override
    public void bindView(LoginView view) {
        mView = view;
        mSubscriptions = new CompositeSubscription();
    }

    @Override
    public void unbindView() {
        mSubscriptions.unsubscribe();
        mView = null;
    }

    @Override
    public void onCreate() {

    }

    @Override
    public void onDestroy() {

    }

    public String getAuthUrl() {
        return AUTH_URL;
    }

    public void loadAccessToken(String code) {
        if (!Utils.getInstance().isNetworkConnected()) {
            mView.errorAuth(R.string.no_network);
            return;
        }

        mView.showProgress();

        Subscription sub = mDataManager.authorization(CLIENT_ID, CLIENT_SECRET, code,
                GRANT_TYPE, REDIRECT_URL)
                .map(authResponseModel -> {
                    Utils.getInstance().writeAuthResponsePref(authResponseModel);
                    return authResponseModel;
                })
                .subscribeOn(mFactory.io())
                .observeOn(mFactory.androidMainThread())
                .subscribe(authResponseModel -> {
                    // Write access token to preferences
                    mView.hideProgress();
                    mView.successAuth();
                }, throwable -> {
                    // If error, then set visible "Sign In" button
                    mView.hideProgress();
                    mView.errorAuth(R.string.error_auth);
                });

        mSubscriptions.add(sub);
    }
}

File: app/src/main/kotlin/com/ne1c/gitteroid/presenters/LoginPresenter.kt
package com.ne1c.gitteroid.presenters

import com.ne1c.gitteroid.R
import com.ne1c.gitteroid.dataproviders.DataManger
import com.ne1c.gitteroid.di.base.ExecutorService
import com.ne1c.gitteroid.di.base.NetworkService
import com.ne1c.gitteroid.ui.views.LoginView
import com.ne1c.rainbowmvp.base.BasePresenter
import rx.subscriptions.CompositeSubscription

class LoginPresenter : BasePresenter<LoginView> {
    companion object {
        val TAG = LoginPresenter::class.java.simpleName

        private val CLIENT_ID = "e94f6920cfbc194174a942fc1a5541355b124309"
        private val CLIENT_SECRET = "bb702baf80daabf7809dc4244f46f37252130c5c"
        private val GRANT_TYPE = "authorization_code"
        private val REDIRECT_URL = "http://about:blank"
        private val RESPONSE_TYPE = "code"

        val AUTH_URL = "https://gitter.im/login/oauth/authorize?client_id=$CLIENT_ID&response_type=$RESPONSE_TYPE&redirect_uri=$REDIRECT_URL"
    }

    private val executor: ExecutorService
    private val dataManager: DataManger
    private val networkService: NetworkService
    private val subscriptions: CompositeSubscription

    constructor(executor: ExecutorService, dataManager: DataManger, networkService: NetworkService) : super() {
        this.executor = executor
        this.dataManager = dataManager
        this.networkService = networkService
        this.subscriptions = CompositeSubscription()
    }

    override fun bindView(view: LoginView?) {
        super.bindView(view)

        if (dataManager.isAuthorize()) {
            mView.successAuth()
        }
    }

    override fun onDestroy() {
        subscriptions.unsubscribe()
    }

    fun loadAccessToken(code: String) {
        if (!networkService.isConnected()) {
            mView?.errorAuth(R.string.no_network)
            return
        }

        mView?.showProgress()

        val sub = dataManager.authorization(CLIENT_ID, CLIENT_SECRET, code, GRANT_TYPE, REDIRECT_URL)
                .map { dataManager.saveUser(it) }
                .subscribeOn(executor.getSubscribeOn())
                .observeOn(executor.getObserveOn())
                .subscribe({
                    mView?.hideProgress()
                    mView?.successAuth()
                }, { throwable ->
                    mView?.hideProgress()
                    mView?.errorAuth(R.string.error_auth)
                })

        subscriptions.add(sub)
    }
}